src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,748 @@
+{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeOperators #-}
 
 {-|
-Module      : Data.BitVector.Sized
+Module      : Data.BitVector.Sized.Internal
 Copyright   : (c) Galois Inc. 2018
 License     : BSD-3
 Maintainer  : benselfridge@galois.com
 Stability   : experimental
 Portability : portable
 
-This module defines a width-parameterized 'BitVector' type and various associated
-operations that assume a 2's complement representation.
+Internal hidden module containing all definitions for the 'BV' type.
 -}
 
 module Data.BitVector.Sized.Internal where
 
-import Data.Bits
-import Data.Ix
-import Data.Parameterized
+import Data.BitVector.Sized.Panic (panic)
+
+import qualified Data.Bits as B
+import qualified Numeric as N
+
+import Data.Char (intToDigit)
+import Data.Int
+import Data.Maybe (fromJust)
+import Data.Word
+import Data.Parameterized ( NatRepr
+                          , natValue
+                          , intValue
+                          , widthVal
+                          , addNat
+                          , ShowF
+                          , EqF(..)
+                          , Hashable(..)
+                          , HashableF(..))
+import qualified Data.Parameterized.NatRepr as P
+import qualified Data.Parameterized.Pair as P
 import GHC.Generics
 import GHC.TypeLits
-import Numeric
-import System.Random
-import Text.PrettyPrint.HughesPJClass
-import Text.Printf
-import Unsafe.Coerce (unsafeCoerce)
+import Language.Haskell.TH.Lift (Lift)
+import Numeric.Natural
+import Prelude hiding (abs, or, and, negate, concat, signum)
+import qualified Prelude as Prelude
+
+----------------------------------------
+-- Utility functions
+
+-- | Check that a 'NatRepr' is representable as an 'Int'.
+checkNatRepr :: NatRepr w -> a -> a
+checkNatRepr = checkNatural . natValue
+
+-- | Check that a 'Natural' is representable as an 'Int'.
+checkNatural :: Natural -> a -> a
+checkNatural n a = if n > fromIntegral (maxBound :: Int)
+  then panic "Data.BitVector.Sized.Internal.checkNatural"
+       [show n <> " not representable as Int"]
+  else a
 
 ----------------------------------------
 -- BitVector data type definitions
 
--- | BitVector datatype, parameterized by width.
-data BitVector (w :: Nat) :: * where
-  BV :: NatRepr w -> Integer -> BitVector w
-  deriving Generic
-
--- | 'BitVector' can be treated as a constructor for pattern matching, but to build
--- one you must use the smart constructor `bitVector`.
-pattern BitVector :: NatRepr w -> Integer -> BitVector w
-pattern BitVector wRepr x <- BV wRepr x
-{-# COMPLETE BitVector #-}
-
--- | Construct a bit vector with a particular width, where the width is inferrable
--- from the type context. The input (an unbounded data type, hence with an
--- infinite-width bit representation), whether positive or negative, is silently
--- truncated to fit into the number of bits demanded by the return type.
+-- | Bitvector datatype, parameterized by width.
+data BV (w :: Nat) :: * where
```

Comment:
Use `Data.Kind.Type` instead of `*` for future-proofness

Reply:
TYVM

Reply:
On that topic, I recommend adding `-Wcompat` to your ghc-options in the cabal file - that has a bunch of warnings that we want to stay on top of to make transition to the next ghc version(s) easier.

Reply:
TYVVM